src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -559,5 +621,529 @@ static void Main()
+class C
+{
+    static void Main()
+    {
+        int x;
+        (x, x) = x;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics(
+                // (7,18): error CS8206: No Deconstruct instance or extension method was found for type 'int'.
+                //         (x, x) = x;
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "x").WithArguments("int").WithLocation(7, 18),
+                // (7,18): error CS0165: Use of unassigned local variable 'x'
+                //         (x, x) = x;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(7, 18)
+                );
+        }
+
+        [Fact]
+        public void Indexers()
+        {
+            string source = @"
+class C
+{
+    static SomeArray array;
+
+    static void Main()
+    {
+        int y;
+        (Foo()[Bar()], y) = new C();
+        System.Console.WriteLine($""Final array values[2] {array.values[2]}"");
+    }
+
+    static SomeArray Foo()
+    {
+        System.Console.WriteLine($""Foo"");
+        array = new SomeArray();
+        return array;
+    }
+
+    static int Bar()
+    {
+        System.Console.WriteLine($""Bar"");
+        return 2;
+    }
+
+    void Deconstruct(out int x, out int y)
+    {
+        System.Console.WriteLine(""Deconstruct"");
+        x = 101;
+        y = 102;
+    }
+}
+class SomeArray
+{
+    public int[] values;
+    public SomeArray() { values = new [] { 42, 43, 44 }; }
+    public int this[int index] {
+        get { System.Console.WriteLine($""indexGet (with value {values[index]})""); return values[index]; }
+        set { System.Console.WriteLine($""indexSet (with value {value})""); values[index] = value; }
+    }
+}
+";
+            var expected =
+@"Foo
+Bar
+Deconstruct
+indexSet (with value 101)
+Final array values[2] 101
+";
+            var comp = CompileAndVerify(source, expectedOutput: expected, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void NullRight()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        int x;
+        (x, x) = null;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics(
+                // (7,9): error CS8210: Deconstruct assignment requires an expression with a type on the right-hand-side.
+                //         (x, x) = null;
+                Diagnostic(ErrorCode.ERR_DeconstructRequiresExpression, "(x, x) = null").WithLocation(7, 9),
+                // (6,13): warning CS0168: The variable 'x' is declared but never used
```

Comment:
So I fixed the BadExpression to contain all the nodes I have from binding and now the warning is this instead:
`// (7,10): error CS0165: Use of unassigned local variable 'x'`
I think this is correct, since BadExpression doesn't really know which nodes are L-values or R-values.